File: 13/soln2.cpp
#include <iostream>
#include <vector>
#include <cmath>

int main(int argc, char** argv)
{
    std::vector<std::pair<unsigned int, int>> pure;
    pure.push_back(std::make_pair(911,0));
    pure.push_back(std::make_pair(827,31));
    pure.push_back(std::make_pair(41,-41));
    pure.push_back(std::make_pair(37,-6));
    pure.push_back(std::make_pair(29,29));
    pure.push_back(std::make_pair(23,23));
    pure.push_back(std::make_pair(19,50));
    pure.push_back(std::make_pair(17,14));
    pure.push_back(std::make_pair(13,13));
    int offset = 41;
    unsigned int increment = 911;
    // pure.push_back(std::make_pair(59,0));
    // pure.push_back(std::make_pair(31,2));
    // pure.push_back(std::make_pair(19,3));
    // pure.push_back(std::make_pair(13,-3));
    // pure.push_back(std::make_pair(7,-4));
    // int offset = 4;
    // unsigned int increment = 59;

    unsigned int print_count = 0;
    unsigned long timestamp = 911 * (std::floor(100000000000000/911) + 1);
    bool success = false;

    while (!success)
    {
        if (print_count++ % 10000000 == 0)
            std::cerr << timestamp << std::endl;

        // check for stop conditions
        success = true;
        for (const auto [t,m] : pure)
        {
            double val = static_cast<double>(timestamp + m) / t;
            // if ((timestamp + m) % t != 0)
            if ((val - std::floor(val)) > 0.00001)
            {
                success = false;
                break;
            }
        }

        if (!success)
            timestamp += increment;
    }
    timestamp -= offset;
    std::cerr << "Success: " << timestamp << std::endl;

}

File: 13/soln2.py
import csv
import copy
import math

if __name__ == "__main__":
    # load data
    data = []
    with open("input1.txt", "r") as csvfile:
        reader = csv.reader(csvfile)
        # skip timestamp; unused
        next(reader)
        times = next(reader)

    # other examples
    # times = [17,"x",13,19]
    # times = [67,7,59,61]
    # times = [67,"x",7,59,61]
    # times = [67,7,"x",59,61]
    # times = [1789,37,47,1889]

    print(times)

    # part #2, iterate through times (at increments of times[0])

    # pre-process times and minutes
    pure_times = []
    pure_minutes = []
    minute = 0
    for t in times:
        if t != "x":
            pure_times.append(int(t))
            pure_minutes.append(minute)
        minute += 1

    # offset by the first number of minutes
    # offset = pure_minutes[pure_times.index(max(pure_times))]
    # pure_minutes = [m - offset for m in pure_minutes]

    # sort from highest to lowest
    pure = zip(pure_times,pure_minutes)
    # pure.sort()
    # pure.reverse()
    print(pure)
    # print(offset)

    # initialize loop variables
    print_count = 0
    timestamp = 100000000000000
    success = False
    increment = pure_times[0]
    while not success:
        print_count += 1
        if print_count % 1000000 == 0:
            print(timestamp)

        # check stop conditions
        success = True
        for t,m in pure:
            if (timestamp + m) % t != 0:
                success = False
                break

        # if we've gotten this far this is the right timestamp
        if not success:
            timestamp += increment

    # include offset
    # timestamp -= offset
    print("Success at time {}".format(timestamp))
